orttraining/orttraining/training_ops/cuda/collective/nccl_kernels.cc

```diff
@@ -43,78 +53,201 @@ Status NcclAllGather::ComputeInternal(OpKernelContext* context) const {
   const size_t element_size = onnx_type->Size();
   ncclDataType_t dtype = GetNcclDataType(onnx_type);
 
-  // Count total number of elements to AllGather.
-  int64_t total_count = 0;
-  for (int i = 0; i < context->InputCount(); i++) {
-    const Tensor* input_tensor = context->Input<Tensor>(i);
-    total_count += input_tensor->Shape().Size();
-  }
-
   // AllGather requires every rank to receive the same amount of data, and
   // slows down significantly if the data is not aligned.  Nvidia recommends 32-byte alignment,
   // so pad to multiple of 32 and world size.
   // Note: the alignment here needs to be kept in-sync with the alignment in zero_optimizer_graph_builder.cc
-  const int64_t alignment = size * 32;
-  const int64_t padded_count = total_count + alignment - (total_count % alignment);
-  const int64_t padded_size = padded_count * element_size;
-  auto fusion_buffer = GetScratchBuffer<void>(padded_size);
-  void* fusion_data = fusion_buffer.get();
+  if (partition_even_) {
+    ORT_ENFORCE(num_input_readies_ == 0);
+    // Count total number of elements to AllGather.
+    int64_t total_count = 0;
+    for (int i = 0; i < context->InputCount(); i++) {
+      const Tensor* input_tensor = context->Input<Tensor>(i);
+      total_count += input_tensor->Shape().Size();
+    }
+    const int64_t alignment = size * 32;
+    const int64_t padded_count = total_count + alignment - (total_count % alignment);
```

Comment:
Haha, good catch, it is not necessary... Jesse initially coded it like this, and since only will be 32 Bytes more in the worst case. We didn't change. Will change it.